future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,472 @@
+    )
+        external
+        authP(ADD_EMPLOYEE_ROLE, arr(initialYearlyDenominationSalary, getTimestamp()))
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            name,
+            getTimestamp()
+        );
+    }
+
+    /**
+     * @dev Add employee to Payroll
+     * @notice Creates employee, adds it to mappings, initializes values.
+               Updates also global Payroll salary sum.
+     * @param accountAddress Employee's address to receive Payroll
+     * @param initialYearlyDenominationSalary Employee's salary
+     * @param name Employee's name
+     * @param startDate It will actually set initial lastPayroll value
+     */
+    function addEmployeeWithNameAndStartDate(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary,
+        string name,
+        uint256 startDate
+    )
+        external
+        authP(ADD_EMPLOYEE_ROLE, arr(initialYearlyDenominationSalary, startDate))
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            name,
+            startDate
+        );
+    }
+
+    /**
+     * @dev Set employee's annual salary
+     * @notice Updates also global Payroll salary sum
+     * @param employeeId Employee's identifier
+     * @param yearlyDenominationSalary Employee's new salary
+     */
+    function setEmployeeSalary(
+        uint128 employeeId,
+        uint256 yearlyDenominationSalary
+    )
+        external
+        authP(ADD_EMPLOYEE_ROLE, arr(yearlyDenominationSalary, 0))
+    {
+        /* check that employee exists */
+        require(employeeIds[employees[employeeId].accountAddress] != 0);
+
+        employees[employeeId].denominationTokenSalary = yearlyDenominationSalary.toSecondDenominationToken();
+    }
+
+    /**
+     * @dev Remove employee from Payroll
+     * @notice Updates also global Payroll salary sum
+     * @param employeeId Employee's identifier
+     */
+    function removeEmployee(uint128 employeeId) external auth(REMOVE_EMPLOYEE_ROLE) {
+        /* check that employee exists */
+        require(employeeIds[employees[employeeId].accountAddress] != 0);
+
+        // Pay owed salary to employee
+        // get time that has gone by (seconds)
+        uint256 time = getTimestamp().sub(employees[employeeId].lastPayroll);
+        if (time > 0)
+            _payTokens(employeeId, time);
+
+        delete employeeIds[employees[employeeId].accountAddress];
+        delete employees[employeeId];
+    }
+
+    /**
+     * @dev Sends ETH to Finance. This contract should never receive funds,
+     *      but in case it happens, this function allows to recover them.
+     * @notice Allows to send ETH from this contract to Finance, to avoid locking them in contract forever.
+     */
+    function escapeHatch() external {
+        // convert ETH to EtherToken
+        etherToken.wrapAndCall.value(this.balance)(address(finance), "Adding Funds");
```

Comment:
I think it makes more sense to move this to the Finance app and remove all ether logic from this contract.

Reply:
So you mean sending ETH to finance? Right now Finance doesn't have a `payable` function. Are you thinking of adding one to wrap ETH and send it to Vault?

Reply:
yes